Factor PDF JS evaluation out of PDFPlugin

hortont424 opened the pull request:
#### 0474d6003056297ae8a98eff1172eae36be468c7
<pre>
Factor PDF JS evaluation out of PDFPlugin
<a href="https://bugs.webkit.org/show_bug.cgi?id=267091">https://bugs.webkit.org/show_bug.cgi?id=267091</a>
<a href="https://rdar.apple.com/118550726">rdar://118550726</a>

Reviewed by Simon Fraser.

Separate PDF JS evaluation code (for auto-printing) out of PDFPlugin.

By virtue of doing this, also enable it in UnifiedPDFPlugin.

And add a test! And make window.print work in MiniBrowser!

* LayoutTests/compositing/plugins/pdf/pdf-plugin-printing-expected.txt: Added.
* LayoutTests/compositing/plugins/pdf/pdf-plugin-printing.html: Added.
* LayoutTests/resources/auto-print.pdf: Added.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::appendValuesInPDFNameSubtreeToVector): Deleted.
(WebKit::getAllValuesInPDFNameTree): Deleted.
(WebKit::getAllScriptsInPDFDocument): Deleted.
(WebKit::jsPDFDocInitialize): Deleted.
(WebKit::jsPDFDocFinalize): Deleted.
(WebKit::PDFPlugin::jsPDFDocClass): Deleted.
(WebKit::PDFPlugin::jsPDFDocPrint): Deleted.
(WebKit::PDFPlugin::makeJSPDFDoc): Deleted.
(WebKit::PDFPlugin::tryRunScriptsInPDFDocument): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::tryRunScriptsInPDFDocument): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::tryRunScriptsInPDFDocument):
(WebKit::PDFPluginBase::print):
Avoid double-running PDF scripts with a &quot;did run&quot; bit (this appears to be broken in shipping Safari).

* Source/WebKit/WebProcess/Plugins/PDF/PDFScriptEvaluator.h: Added.
(WebKit::PDFScriptEvaluator::PDFScriptEvaluator):
(WebKit::PDFScriptEvaluator::create):
* Source/WebKit/WebProcess/Plugins/PDF/PDFScriptEvaluator.mm: Added.
(WebKit::appendValuesInPDFNameSubtreeToVector):
(WebKit::getAllValuesInPDFNameTree):
(WebKit::getAllScriptsInPDFDocument):
(WebKit::jsPDFDocInitialize):
(WebKit::jsPDFDocFinalize):
(WebKit::PDFScriptEvaluator::jsPDFDocClass):
(WebKit::PDFScriptEvaluator::jsPDFDocPrint):
(WebKit::PDFScriptEvaluator::runScripts):
(WebKit::PDFScriptEvaluator::print):
Factor this code out of PDFPlugin.

* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteHost.mm:
(WebKit::PlatformCALayerRemoteHost::create):
Unified sources fixes.

* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController _webView:printFrame:pdfFirstPageSize:completionHandler:]):
Hook up the printing UI delegate in MiniBrowser.

Canonical link: <a href="https://commits.webkit.org/272706@main">https://commits.webkit.org/272706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d29e3f4fdc18a6a120dc7f7351614dd38f66852

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32781 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11532 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34641 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/35397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/29614 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33708 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13876 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8717 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/35397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33224 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9725 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/29274 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/35397 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8607 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29226 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36697 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29782 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29632 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34724 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/8728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6694 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32590 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10413 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/28900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7613 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9338 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9343 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
LGTM

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+
+namespace WebKit {
+
+static void appendValuesInPDFNameSubtreeToVector(CGPDFDictionaryRef subtree, Vector<CGPDFObjectRef>& values)
```
Should this take a `Vector<RetainPtr<CGPDFObjectRef>>`?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+    appendValuesInPDFNameSubtreeToVector(tree, allValues);
+}
+
+static void getAllScriptsInPDFDocument(CGPDFDocumentRef pdfDocument, Vector<RetainPtr<CFStringRef>>& scripts)
```
Maybe the argument should be a `RetainPtr<CGPDFDocumentRef>` ?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+    getAllValuesInPDFNameTree(javaScriptNameTree, objects);
+    size_t objectCount = objects.size();
+
+    for (size_t i = 0; i < objectCount; ++i) {
```
Modern loop?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+
+    for (size_t i = 0; i < objectCount; ++i) {
+        CGPDFDictionaryRef javaScriptAction;
+        if (!CGPDFObjectGetValue(reinterpret_cast<CGPDFObjectRef>(objects[i]), kCGPDFObjectTypeDictionary, &javaScriptAction))
```
Why the `reinterpret_cast<>` ?

hortont424 replied:
a grand question

hortont424 replied:
removed

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+        CFIndex length = 0;
+        CGPDFStreamRef stream;
+        CGPDFStringRef string;
+        RetainPtr<CFDataRef> data;
```
This can move inside the condition

smfr replied:
Or not, since you want to ensure that `bytes` remains value. Would be nice to rewrite this to avoid a dangling `bytes` value (maybe with a lambda).

hortont424 replied:
Rewrote all of this

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+        if (!bytes || !length)
+            continue;
+
+        CFStringEncoding encoding = (length > 1 && bytes[0] == 0xFE && bytes[1] == 0xFF) ? kCFStringEncodingUnicode : kCFStringEncodingUTF8;
```
There must be some better way to do this.

hortont424 replied:
No doubt, but this one I'm going to leave alone because any alternative would probably involve creating some needless WTF type here.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+
+static void jsPDFDocInitialize(JSContextRef ctx, JSObjectRef object)
+{
+    PDFPluginScriptEvaluator* evaluator = static_cast<PDFPluginScriptEvaluator*>(JSObjectGetPrivate(object));
```
Can we make it possible to `dynamicDowncast<>`?

hortont424 replied:
from void*? I don't think so

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+    return JSValueMakeUndefined(ctx);
+}
+
+void PDFPluginScriptEvaluator::runScripts(CGPDFDocumentRef document, Client& client)
```
I think we should ensure that we keep the `CGPDFDocumentRef` alive for the duration of this function.

hortont424 replied:
PDFPluginScriptEvaluator keeps a strong reference to it.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+{
+    ASSERT(isMainRunLoop());
+
+    Ref<PDFPluginScriptEvaluator> evaluator = PDFPluginScriptEvaluator::create(document, client);
```
`auto`

hortont424 replied:
Actually shouldn't it be `Ref evaluator` now?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginScriptEvaluator.mm:
```diff
@@ -0,0 +1,220 @@
+            return;
+
+        JSGlobalContextRef ctx = JSGlobalContextCreate(nullptr);
+        JSObjectRef jsPDFDoc = JSObjectMake(ctx, jsPDFDocClass(), evaluator.ptr());
```
Does this need releasing?

hortont424 replied:
this I cannot puzzle out, but other cases of JSObjectMake suggest no

hortont424 committed:
Factor PDF JS evaluation out of PDFPlugin
https://bugs.webkit.org/show_bug.cgi?id=267091
rdar://118550726

Reviewed by Simon Fraser.

Separate PDF JS evaluation code (for auto-printing) out of PDFPlugin.

By virtue of doing this, also enable it in UnifiedPDFPlugin.

And add a test! And make window.print work in MiniBrowser!

* LayoutTests/compositing/plugins/pdf/pdf-plugin-printing-expected.txt: Added.
* LayoutTests/compositing/plugins/pdf/pdf-plugin-printing.html: Added.
* LayoutTests/resources/auto-print.pdf: Added.
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::appendValuesInPDFNameSubtreeToVector): Deleted.
(WebKit::getAllValuesInPDFNameTree): Deleted.
(WebKit::getAllScriptsInPDFDocument): Deleted.
(WebKit::jsPDFDocInitialize): Deleted.
(WebKit::jsPDFDocFinalize): Deleted.
(WebKit::PDFPlugin::jsPDFDocClass): Deleted.
(WebKit::PDFPlugin::jsPDFDocPrint): Deleted.
(WebKit::PDFPlugin::makeJSPDFDoc): Deleted.
(WebKit::PDFPlugin::tryRunScriptsInPDFDocument): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::tryRunScriptsInPDFDocument): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::tryRunScriptsInPDFDocument):
(WebKit::PDFPluginBase::print):
Avoid double-running PDF scripts with a "did run" bit (this appears to be broken in shipping Safari).

* Source/WebKit/WebProcess/Plugins/PDF/PDFScriptEvaluator.h: Added.
(WebKit::PDFScriptEvaluator::PDFScriptEvaluator):
(WebKit::PDFScriptEvaluator::create):
* Source/WebKit/WebProcess/Plugins/PDF/PDFScriptEvaluator.mm: Added.
(WebKit::appendValuesInPDFNameSubtreeToVector):
(WebKit::getAllValuesInPDFNameTree):
(WebKit::getAllScriptsInPDFDocument):
(WebKit::jsPDFDocInitialize):
(WebKit::jsPDFDocFinalize):
(WebKit::PDFScriptEvaluator::jsPDFDocClass):
(WebKit::PDFScriptEvaluator::jsPDFDocPrint):
(WebKit::PDFScriptEvaluator::runScripts):
(WebKit::PDFScriptEvaluator::print):
Factor this code out of PDFPlugin.

* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteHost.mm:
(WebKit::PlatformCALayerRemoteHost::create):
Unified sources fixes.

* Tools/MiniBrowser/mac/WK2BrowserWindowController.m:
(-[WK2BrowserWindowController _webView:printFrame:pdfFirstPageSize:completionHandler:]):
Hook up the printing UI delegate in MiniBrowser.

Canonical link: https://commits.webkit.org/272706@main
